meta-facebook/yv35-cl/src/platform/plat_mctp.c

```diff
@@ -0,0 +1,286 @@
+#define MCTP_IC_MASK 0x80
+
+/* i3c 8-bit addr */
+#define I3C_STATIC_ADDR_BIC		0x40
+#define I3C_STATIC_ADDR_BMC		0x20
+
+/* i3c dev bus */
+#define I3C_BUS_BMC	0
+
+/* mctp endpoint */
+#define MCTP_EID_BMC 0x01
+#define MCTP_EID_SELF 0x02
+
+K_TIMER_DEFINE(send_cmd_timer, send_cmd_to_dev, NULL);
+K_WORK_DEFINE(send_cmd_work, send_cmd_to_dev_handler);
+
+typedef struct _mctp_smbus_port {
+	mctp *mctp_inst;
+	mctp_medium_conf conf;
+	uint8_t user_idx;
+} mctp_smbus_port;
+
+typedef struct _mctp_i3c_port {
+	mctp *mctp_inst;
+	mctp_medium_conf conf;
+	uint8_t user_idx;
+} mctp_i3c_port;
+
+/* mctp route entry struct */
+typedef struct _mctp_route_entry {
+	uint8_t endpoint;
+	uint8_t bus; /* TODO: only consider smbus/i3c */
+	uint8_t addr; /* TODO: only consider smbus/i3c */
+} mctp_route_entry;
+
+typedef struct _mctp_msg_handler {
+	MCTP_MSG_TYPE type;
+	mctp_fn_cb msg_handler_cb;
+} mctp_msg_handler;
+
+static mctp_i3c_port i3c_port[] = {
+	{ .conf.i3c_conf.bus = I3C_BUS_BMC, .conf.i3c_conf.addr = I3C_STATIC_ADDR_BMC},
+};
+
+mctp_route_entry mctp_route_tbl[] = {
+	{ MCTP_EID_BMC, I3C_BUS_BMC, I3C_STATIC_ADDR_BMC },
+};
+
+static mctp *find_mctp_by_i3c(uint8_t bus)
+{
+	uint8_t i;
+	for (i = 0; i < ARRAY_SIZE(i3c_port); i++) {
+		mctp_i3c_port *p = i3c_port + i;
+
+		if (bus == p->conf.i3c_conf.bus) {
+			return p->mctp_inst;
+		}
+	}
+
+	return NULL;
+}
+
+static void set_endpoint_resp_handler(void *args, uint8_t *buf, uint16_t len)
+{
+	CHECK_NULL_ARG(buf);
+
+	if (len < 0) {
+		return;
+	}
+
+	LOG_HEXDUMP_DBG(buf, len, __func__);
+}
+
+static void set_endpoint_resp_timeout(void *args)
+{
+	mctp_route_entry *p = (mctp_route_entry *)args;
+	LOG_DBG("[%s] Endpoint 0x%x set endpoint failed on bus %d \n", __func__, p->endpoint,
+	       p->bus);
+}
+
+static void set_dev_endpoint(void)
+{
+	for (uint8_t i = 0; i < ARRAY_SIZE(mctp_route_tbl); i++) {
+		mctp_route_entry *p = mctp_route_tbl + i;
+
+		/* skip BMC */
+		if (p->bus == I3C_BUS_BMC && p->addr == I3C_STATIC_ADDR_BMC)
+			continue;
+
+		for (uint8_t j = 0; j < ARRAY_SIZE(i3c_port); j++) {
+			if (p->bus != i3c_port[j].conf.i3c_conf.bus)
+				continue;
+
+			struct _set_eid_req req = { 0 };
+			req.op = SET_EID_REQ_OP_SET_EID;
+			req.eid = p->endpoint;
+
+			mctp_ctrl_msg msg;
+			memset(&msg, 0, sizeof(msg));
+			msg.ext_params.type = MCTP_MEDIUM_TYPE_SMBUS;
+			msg.ext_params.smbus_ext_params.addr = p->addr;
+
+			msg.hdr.cmd = MCTP_CTRL_CMD_SET_ENDPOINT_ID;
+			msg.hdr.rq = 1;
+
+			msg.cmd_data = (uint8_t *)&req;
+			msg.cmd_data_len = sizeof(req);
+
+			msg.recv_resp_cb_fn = set_endpoint_resp_handler;
+			msg.timeout_cb_fn = set_endpoint_resp_timeout;
+			msg.timeout_cb_fn_args = p;
+
+			mctp_ctrl_send_msg(find_mctp_by_i3c(p->bus), &msg);
+		}
+	}
+}
+
+static uint8_t mctp_msg_recv(void *mctp_p, uint8_t *buf, uint32_t len, mctp_ext_params ext_params)
+{
+	CHECK_NULL_ARG_WITH_RETURN(mctp_p, MCTP_ERROR);
+	CHECK_NULL_ARG_WITH_RETURN(buf, MCTP_ERROR);
+
+	if (len < 0)
```

Comment:
Same as above...len is defined as uint32_t so this check is redundant.

Reply:
Remove it, thank you.